packages/model-viewer/src/features/annotation.ts

```diff
@@ -0,0 +1,208 @@
+    super(wrapper);
+    this.normal = new Vector3();
+    this.updatePosition(element);
+    this.updateNormal(element);
+  }
+
+  updatePosition(element: HTMLElement) {
+    const positionNodes = parseExpressions(element.dataset.position!)[0].terms;
+    for (let i = 0; i < 3; ++i) {
+      this.position.setComponent(
+          i, normalizeUnit(positionNodes[i] as NumberNode<'m'>).number);
+    }
+  }
+
+  updateNormal(element: HTMLElement) {
+    const normalNodes = parseExpressions(element.dataset.normal!)[0].terms;
+    for (let i = 0; i < 3; ++i) {
+      this.normal.setComponent(
+          i, normalizeUnit(normalNodes[i] as NumberNode<'m'>).number);
+    }
+  }
+}
+
+export declare interface AnnotationInterface {
+  hiddenOpacity: number;
+  hiddenAngle: string;
+  addHotspot(element: HTMLElement): void;
+}
+
+export const AnnotationMixin = <T extends Constructor<ModelViewerElementBase>>(
+    ModelViewerElement: T): Constructor<AnnotationInterface>&T => {
+  class AnnotationModelViewerElement extends ModelViewerElement {
+    @property({type: Number, attribute: 'hidden-opacity'})
+    hiddenOpacity: number = DEFAULT_HIDDEN_OPACITY;
+
+    @style({
+      intrinsics: {
+        basis: [numberNode(DEFAULT_HIDDEN_ANGLE, 'rad')],
+        keywords: {auto: [null]}
+      },
+      updateHandler: $syncHiddenAngle
+    })
+    @property({type: String, attribute: 'hidden-angle'})
+    hiddenAngle: string = 'auto';
+
+    private[$annotationRenderer] = new CSS2DRenderer();
+    private[$hiddenAngle] = DEFAULT_HIDDEN_ANGLE;
+    private[$nextHotspotIndex] = 0;
+    private[$hotspotMap] = new Map();
+    private[$observer] = new MutationObserver((mutations) => {
+      mutations.forEach((mutation) => {
+        switch (mutation.type) {
+          case 'childList':
+            mutation.addedNodes.forEach((node) => {
+              this.addHotspot(node);
+            });
+            mutation.removedNodes.forEach((node) => {
+              this.removeHotspot(node);
+            });
+            break;
+          case 'attributes':
+            const {target} = mutation;
+            const hotspot = this[$hotspotMap].get(target);
+            if (!hotspot) {
+              this.addHotspot(target);
+              break;
+            }
+            switch (mutation.attributeName) {
+              case 'data-position':
+                hotspot.updatePosition(target);
+                break;
+              case 'data-normal':
+                hotspot.updateNormal(target);
+                break;
+            }
+            break;
+        }
+      });
+    });
+
+    connectedCallback() {
+      super.connectedCallback();
+      const {domElement} = this[$annotationRenderer];
+      const {style} = domElement;
+      style.pointerEvents = 'none';
+      style.position = 'absolute';
+      style.top = '0';
+      this.shadowRoot!.querySelector('.container')!.appendChild(domElement);
+
+      for (let i = 0; i < this.children.length; ++i) {
```

Comment:
This approach will attempt to add every child, including nodes that are possibly not intended to be hotspots, as hotspots. For example, consider this configuration:

```html
<model-viewer>
  <img slot="poster" src="foo.png" data-position="..." data-normal="...">
</model-viewer>
```

Now, it is fair to say that the user annotating this node as though it were a hotspot and then slotting it as the poster is a bit fishy. However, the point is that we are _inferring_ that the user meant to create a hotspot when, in fact, the user's intention is ambiguous. `data-*` attributes are, after all, intended for authors to decorate their nodes in any way for any reason.

For a declarative API, I would recommend that we have an explicit signal. For this change, we could punt on that and focus on the imperative API. OTOH if we do it now I would probably suggest a more explicit signal. For example, the author could pre-populate the slot attribute:

```html
<model-viewer>
  <div id="some-hotspot" slot="hotspot-foo" data-position="..." data-normal="..."></div>
</model-viewer>
```

Then, we would create hotspots out of all nodes that have a `slot="hotspot-*"` attribute.

Reply:
Fixed.